src/Database/LSMTree/Internal/BlobRef.hs

```diff
@@ -4,185 +4,199 @@
-import qualified Data.Primitive.ByteArray as P (MutableByteArray,
-                     newPinnedByteArray, unsafeFreezeByteArray)
+import qualified Data.Primitive.ByteArray as P (newPinnedByteArray,
+                     unsafeFreezeByteArray)
 import qualified Data.Vector as V
-import           Data.Word (Word32, Word64)
+import qualified Data.Vector.Mutable as VM
+import           Database.LSMTree.Internal.BlobFile (BlobFile (..),
+                     BlobSpan (..))
+import qualified Database.LSMTree.Internal.BlobFile as BlobFile
 import qualified Database.LSMTree.Internal.RawBytes as RB
 import           Database.LSMTree.Internal.Serialise (SerialisedBlob (..))
 import qualified System.FS.API as FS
 import           System.FS.API (HasFS)
 import qualified System.FS.BlockIO.API as FS
+import           System.FS.BlockIO.API (HasBlockIO)
 
--- | A handle-like reference to an on-disk blob. The blob can be retrieved based
--- on the reference.
+
+-- | A raw blob reference is a reference to a blob within a blob file.
 --
--- See 'Database.LSMTree.Common.BlobRef' for more info.
-data BlobRef m h = BlobRef {
-      blobRefFile  :: !h
-    , blobRefCount :: {-# UNPACK #-} !(RefCounter m)
-    , blobRefSpan  :: {-# UNPACK #-} !BlobSpan
+-- The \"raw\" means that it does no reference counting, so does not maintain
+-- ownership of the 'BlobFile'. Thus these are only safe to use in the context
+-- of code that already (directly or indirectly) owns the blob file that the
+-- blob ref uses (such as within run merging).
+--
+-- Thus these cannot be handed out via the API. Use 'WeakBlobRef' for that.
+--
+data RawBlobRef m h = RawBlobRef {
+      rawBlobRefFile :: {-# NOUNPACK #-} !(BlobFile m h)
+    , rawBlobRefSpan :: {-# UNPACK #-}   !BlobSpan
     }
   deriving stock (Show)
 
-instance NFData h => NFData (BlobRef m h) where
-  rnf (BlobRef a b c) = rnf a `seq` rnf b `seq` rnf c
-
--- | Location of a blob inside a blob file.
-data BlobSpan = BlobSpan {
-    blobSpanOffset :: {-# UNPACK #-} !Word64
-  , blobSpanSize   :: {-# UNPACK #-} !Word32
-  }
-  deriving stock (Show, Eq)
-
-instance NFData BlobSpan where
-  rnf (BlobSpan a b) = rnf a `seq` rnf b
-
-blobRefSpanSize :: BlobRef m h -> Int
-blobRefSpanSize = fromIntegral . blobSpanSize . blobRefSpan
+-- | A \"weak\" reference to a blob within a blob file. These are the ones we
+-- can return in the public API and can outlive their parent table.
+--
+-- They are weak references in that they do not keep the file open using a
+-- reference count. So when we want to use our weak reference we have to
+-- dereference them to obtain a normal strong reference while we do the I\/O
+-- to read the blob. This ensures the file is not closed under our feet.
+--
+-- See 'Database.LSMTree.Common.BlobRef' for more info.
+--
+data WeakBlobRef m h = WeakBlobRef {
+      weakBlobRefFile :: {-# NOUNPACK #-} !(BlobFile m h)
+    , weakBlobRefSpan :: {-# UNPACK #-}   !BlobSpan
+    }
+  deriving stock (Show)
```

Comment:
Is there any difference between raw and weak blob references, other than that the names serve as documentation? Is it necessary to make the distinction if both are a blob reference without ownership?

I'm probably answering my own question here, but yes the dinstinction is necessary, because weak references have to be upgraded before reading the blobs. However, you could make the same distinction by getting rid of `RawBlobRef` and providing `unsafeReadWeakBlobRef` and `readWeakBlobRef` functions

Reply:
Weak ones will change to use `WeakRef` once we use the new `Ref` API.

A raw one is "safe" to read, but you must already own a reference. A weak one cannot be read without upgrading to a strong ref.